src/test_fixture/transport/routing.rs

```diff
@@ -0,0 +1,214 @@
+        write!(f, "Subscribe[{:?}]", self.subscription)
+    }
+}
+
+impl SubscribeRequest {
+    pub fn new(
+        subscription: SubscriptionType,
+        link: mpsc::Sender<CommandEnvelope>,
+    ) -> (Self, oneshot::Receiver<()>) {
+        let (ack_tx, ack_rx) = oneshot::channel();
+        (
+            Self {
+                subscription,
+                link,
+                ack_tx,
+            },
+            ack_rx,
+        )
+    }
+
+    pub fn acknowledge(self) {
+        self.ack_tx.send(()).unwrap();
+    }
+
+    pub fn subscription(&self) -> SubscriptionType {
+        self.subscription
+    }
+
+    pub fn sender(&self) -> mpsc::Sender<CommandEnvelope> {
+        self.link.clone()
+    }
+}
+
+/// Switch state
+#[derive(Debug)]
+enum State {
+    /// Getting ready to start receiving commands. In this state, it is possible to add new
+    /// peer connections
+    Idle(
+        mpsc::Receiver<SwitchCommand>,
+        HashMap<HelperIdentity, mpsc::Receiver<TransportCommand>>,
+    ),
+    /// Interim state where switch is about to start actively listening for incoming traffic
+    Preparing,
+    /// Actively listening. It is no longer possible to change the state of this switch and add
+    /// or remove connections.
+    Listening(JoinHandle<()>),
+}
+
+/// Takes care of forwarding commands received from multiple links (one link per peer)
+/// to the subscribers
+pub(super) struct Switch {
+    state: State,
+    tx: mpsc::Sender<SwitchCommand>,
+    id: HelperIdentity,
+}
+
+impl Debug for Switch {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        write!(f, "switch[{:?}]", self.state)
+    }
+}
+
+impl Switch {
+    pub fn new(id: HelperIdentity) -> Self {
+        let (tx, rx) = mpsc::channel(1);
+
+        Self {
+            state: State::Idle(rx, HashMap::default()),
+            tx,
+            id,
+        }
+    }
+
+    pub fn new_peer(&mut self, peer_id: HelperIdentity, peer_rx: mpsc::Receiver<TransportCommand>) {
+        let State::Idle(_, peers) = &mut self.state else {
+            panic!("Not in Idle state");
+        };
+
+        assert!(peers.insert(peer_id, peer_rx).is_none());
+    }
+
+    /// Starts listening to the incoming messages in a separate task. Can only be called once
+    /// and only when it is in the `Idle` state.
+    pub fn listen(&mut self) {
+        let State::Idle(mut rx, peers) = std::mem::replace(&mut self.state, State::Preparing) else {
+            panic!("Not in Idle state");
+        };
+
+        let mut peer_links = SelectAll::new();
+        for (addr, link) in peers {
+            peer_links.push(ReceiverStream::new(link).map(move |command| (addr.clone(), command)));
+        }
+        let handle = tokio::spawn(async move {
+            let mut query_router = QueryCommandRouter::default();
+            loop {
+                ::tokio::select! {
+                    Some(command) = rx.recv() => {
+                        match command {
```

Comment:
very likely yes, I would probably postpone that change till I implement query management commands - would have a more clear picture how it would look like